third_party/java/java_tools/update_java_tools.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/bash -e
+#
+# A script to update the checked-in jars corresponding to Java tools used
+# by the Java rules in Bazel.
+#
+# For usage please run
+# ~/third_party/java/java_tools/update_java_tools.sh help
+
+# Maps the java tool names to their associated bazel target.
+declare -A tool_name_to_target=(["JavaBuilder"]="src/java_tools/buildjar:JavaBuilder_deploy.jar" \
+["VanillaJavaBuilder"]="src/java_tools/buildjar:VanillaJavaBuilder_deploy.jar" \
+["GenClass"]="src/java_tools/buildjar/java/com/google/devtools/build/buildjar/genclass:GenClass_deploy.jar" \
+["Runner"]="src/java_tools/junitrunner/java/com/google/testing/junit/runner:Runner_deploy.jar" \
+["ExperimentalRunner"]="src/java_tools/junitrunner/java/com/google/testing/junit/runner:ExperimentalRunner_deploy.jar" \
+["JacocoCoverage"]="src/java_tools/junitrunner/java/com/google/testing/coverage:JacocoCoverage_jarjar_deploy.jar" \
+["Turbine"]="src/java_tools/buildjar/java/com/google/devtools/build/java/turbine/javac:turbine_deploy.jar" \
+["TurbineDirect"]="src/java_tools/buildjar/java/com/google/devtools/build/java/turbine:turbine_direct_binary_deploy.jar" \
+["SingleJar"]="src/java_tools/singlejar/java/com/google/devtools/build/singlejar:bazel-singlejar_deploy.jar")
+
+usage="This script updates the checked-in jars corresponding to the tools "\
+"used by the Java rules in Bazel.
+
+To update all the tools simultaneously run from your bazel workspace root:
+~/third_party/java/java_tools/update_java_tools.sh
+
+To update only one or one subset of the tools run
+third_party/java/java_tools/update_java_tools.sh tool_1 tool_2 ... tool_n
+
+where tool_i can have one of the values: ${!tool_name_to_target[*]}.
+
+For example, to update only JavaBuilder run
+third_party/java/java_tools/update_java_tools.sh JavaBuilder
+
+To update JavaBuilder, Turbine and SingleJar run
+third_party/java/java_tools/update_java_tools.sh JavaBuilder Turbine SingleJar
+"
+
+if [[ ! -z "$1" && $1 = "help" ]]; then
+  echo "$usage"
+  exit
+fi
+
+# Stores the names of the tools required for update.
+tools_to_update=()
+if [[ ! -z "$@" ]]
+then
+   # Update only the tools specified on the command line.
+   tools_to_update=("$@")
+else
+  # If no tools were specified update all of them.
+  tools_to_update=(${!tool_name_to_target[*]})
+fi
+
+# Stores the workspace relative path of all the tools that were updated
+# (e.g. third_party/java/java_tools/JavaBuilder_deploy.jar)
+updated_tools=()
+
+# Updates the tool with the given bazel target.
+#
+# Builds the given bazel target and copies the generated binary
+# (which can be either under bazel-bin/ or bazel-genfiles/) under
+# third_party/java/java_tools.
+#
+# Fails if the bazel build fails.
+#
+# bazel_target   The target to be built with bazel.
+# tool_name      The name of the tool associated with the given bazel
+#                target. Used only for printing error messages.
+function update_tool() {
+  local bazel_target="${1}"; shift
```

Comment:
nit: `"$1"` is enough -- there isn't any need to disambiguate where the name of the variable ends.